### ISSUE TITLE
Task / issue / user story dependencies

### ISSUE THREAD

**Issue by nidico** [taigaio/taiga-back#154]
There's one strictly needed feature which prevents us from using Taiga: The lack of task (and also issue / user story) dependencies.

The required functionality is:
- Allow to connect tasks and apply semantics to the connection (directed dependency / undirected relation; Redmine has a couple more, but these are the important ones)
- Expose in UX, e.g.
  - Minimum: Show "blocked by xy" on taskcard
  - Unhighlight tasks which are blocked yet
  - Color-highlight related / blocking tasks in the Kanban board on hover (or allow to select tasks on the Kanban board)
  - Prevent tasks from being moved if it depends on a task which isn't in that state yet.
  - Same goes for issues, user stories. Should also be interrelatable.
  - UX designers will have more ideas...

Note: There was already an issue (#5) on that topic which has been closed.

**Comment by Zorlin** [comment]
Consider this my very, very strong +1.

@superalex We are trying to use Taiga for systems administration, and in particular a very long set of evolving and complex dependencies that need to be solved to improve our architecture. We tried using Redmine for this in the past but it was horrible for various reasons - too many clicks, too many page loads and pretty much a million other reasons.

Taiga is SO CLOSE to perfect for us, though! We literally just need:
* The ability to nest task dependencies (A is dependent on B which is dependent on C which is dependent on D).
* The ability to do dependencies in Issues (this is ridiculously critical for us long term - to the point where I'm actually going to be implementing this in the next week and literally just pasting "Depends on #NUM" in every single task).

Want-to-haves:
* The ability to customize: TYPE, STATUS, SEVERITY, PRIORITIES
* The ability to customize the user story points - we have no use for the existing ones as we do not do application/web development, but we do have a lot of things we could replace them with.

As an example, if I had the ability to customize the points, I would use "Risk", "Simplify", "Depends" and "Perf" as those are the most important four for us. If I could go further and add a few more points categories, I would add "HA" or "Reliable", "Automate" and others.

I can accomplish much of what I want by simply hacking around the limitations that Taiga currently has (AFAIK I can just edit the JSON file in taiga/projects/fixtures/initial_project_templates.json prior to creating a project - would I be able to get away with totally editing "UX" *including* the slug?) and I'm extremely excited by that - so regardless of whether or not Taiga actually implements this stuff, I think this is the first tool I've found that will actually do what our team needs (a HUGE improvement over what we're using now, at the very least). But please consider our use case! For extended, long-term sysadmin projects these sorts of features could end up being literally make or break.

The want-to-haves are both completely irrelevant to this issue but provide some context for what I'm envisioning and our overall use case. Thanks for any consideration.

**Comment by superalex** [comment]
Hello Benjamin,

Thank you very much for using Taiga!. we are very happy that Taiga is so
close for you, I hope it can finally fit :)

Implementing dependencies is in our roadmap but we would like to develop
some features before and so I can't give you an aprox date for it.

Customization of type, status, severity, priorities and story points is
something you can do using right now. Admin -> custom attributes.

The json files you talk about are the two templates by default for Taiga
(kanban and backlog) you have more info in this github issue ->
https://github.com/taigaio/taiga-front/issues/234

Regards!,

2015-01-13 12:08 GMT+01:00 Benjamin Arntzen <notifications@github.com>:

> Consider this my very, very strong +1.
>
> @superalex <https://github.com/superalex> We are trying to use Taiga for
> systems administration, and in particular a very long set of evolving and
> complex dependencies that need to be solved to improve our architecture. We
> tried using Redmine for this in the past but it was horrible for various
> reasons - too many clicks, too many page loads and pretty much a million
> other reasons.
>
> Taiga is SO CLOSE to perfect for us, though! We literally just need:
>
>    - The ability to nest task dependencies (A is dependent on B which is
>    dependent on C which is dependent on D).
>    - The ability to do dependencies in Issues (this is ridiculously
>    critical for us long term - to the point where I'm actually going to be
>    implementing this in the next week and literally just pasting "Depends on
>    #NUM" in every single task).
>
> Want-to-haves:
>
>    - The ability to customize: TYPE, STATUS, SEVERITY, PRIORITIES
>    - The ability to customize the user story points - we have no use for
>    the existing ones as we do not do application/web development, but we do
>    have a lot of things we could replace them with.
>
> As an example, if I had the ability to customize the points, I would use
> "Risk", "Simplify", "Depends" and "Perf" as those are the most important
> four for us. If I could go further and add a few more points categories, I
> would add "HA" or "Reliable", "Automate" and others.
>
> I can accomplish much of what I want by simply hacking around the
> limitations that Taiga currently has (AFAIK I can just edit the JSON file
> in taiga/projects/fixtures/initial_project_templates.json prior to creating
> a project - would I be able to get away with totally editing "UX"
> *including* the slug?) and I'm extremely excited by that - so regardless
> of whether or not Taiga actually implements this stuff, I think this is the
> first tool I've found that will actually do what our team needs (a HUGE
> improvement over what we're using now, at the very least). But please
> consider our use case! For extended, long-term sysadmin projects these
> sorts of features could end up being literally make or break.
>
> The want-to-haves are both completely irrelevant to this issue but provide
> some context for what I'm envisioning and our overall use case. Thanks for
> any consideration.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/taigaio/taiga-back/issues/154#issuecomment-69728991>.
>



-- 

Alejandro Alonso Fernández
http://alejandro.alonsofernandez.me
<http://t.signaletre.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XZsfDd1sW3MPfyH4Wr-ZRW1p1hXH56dzNPf4jXYzd02?t=http%3A%2F%2Falejandro.alonsofernandez.me%2F&si=5071763271057408&pi=9c62b84b-783f-48ea-bad0-43fdb75c033d>

**Comment by Zorlin** [comment]
@superalex Awesome, thank you! That helps a lot. The only thing missing there is editing the categories to assign to points unless I'm missing something - I'd like to change "UX", "Front", " Back" to other stuff. The actual amount of points selectable is totally fine right now :)

**Comment by Zorlin** [comment]
We can live without dependencies for maybe 1-3 months or a little longer if we have to... Taiga will still be much better than what we're using. Would it still be helpful for me to provide a chart of how we're using dependencies?

**Comment by superalex** [comment]
Sure @Zorlin it would be really helpful!

**Comment by Zorlin** [comment]
@superalex - awesome. I'll try to get that to you guys as soon as we have it which might be a while. We're usually in firefighting mode which we want to get out of by finding project management stuff we enjoy.

As a sidenote, kudos for thinking of this stuff before I suggested it. The customization options are a little hard to surface, is the only thing, but most projects don't allow this level of customization until they're much older (in open-source years).

As a sidenote, I think it might be cool to have user guides on customization eventually. I'd be happy to write one once we've started really hitting our stride, maybe with examples of how we made it fit our workflow.

Anyways, sorry to veer off topic so badly :)

**Comment by superalex** [comment]
You are right, that customizations can be a bit hard to find and we would have to write/record/show any customization guides, if you document write any of them with your own examples it would be great, I'm sure a lot of Taiga users would find it really useful :+1: 

Regards!,

**Comment by nehalecky** [comment]
+1 dependancies

**Comment by ncoghlan** [comment]
Hi @superalex, thanks for the Taiga team's work bringing compelling open source project management tools to the world!

Is there any update on the dependency management discussions and where they might fit on the current road map?

**Comment by superalex** [comment]
Hello @ncoghlan,

Sorry but we have no news about dependency management.

Regards!,

**Comment by kafoso** [comment]
Relations between tasks and issues is a must-have in a project management tool. Otherwise, how is it different compared to having a To-do list (e.g. in an application) and GitHubs very fine and simple Issue system?

A relation should create a two-way link between - let's call them "entries" for ease of reference; "entries" being tasks, issues, User Stories, Epics, etc.

Consider the case:

1. In **Entry A**, a user creates a link/relation to **Entry B**.
2. Now, whenever user goes to **Entry B**, the user sees a link pointing back to **Entry A**.

Currently in Taiga, the only means of making relationships between entries are:

1. In a User Story, create tasks under the "Related tasks" heading. "Related" is horrible naming, by the way. Rather, they're "Dependencies" or simply "Sub-tasks".
2. Go to **Entry A** and put in the description that it has a relation to **Entry B**. Now go to **Entry B** and repeat the previous step, but instead linking to **Entry A**. It's redundant and annoying.

Moreover, a relation ship can fall in several categories or types. It's not just a two-headed arrow.

I'm sure you're away of many of JIRA's features. Their link logic with "Related", "Blocks", "Blocked by", etc. is down right excellent. I would love to see that integrated into Taiga.

Additionally, a link structure as in JIRA would to some extend leave out the need for a hierarchical task structure, since tasks can be specified to depend on each other (through "Blocks" and "Blocked by").

**Comment by fcouceiro** [comment]
@superalex I think you can "resolve" this by letting custom fields reference tasks with markdown.

That way one can create a custom field "Depends on" or "Blocked by" or another that fits its needs and simply reference the task/US/issue as normal (#178, for instance).

Regards

**Comment by superalex** [comment]
Thanks for your suggestions @fcouceiro

**Comment by oalexandersson** [comment]
Hi!

We just started using Taiga, and love how lean it feels compared to JIRA, for example. Story/task/issue links, however, is a must-have for us to be able to continue using Taiga. :disappointed:

First, I would like to comment on @fcouceiro 's suggestion: This wouldn't result in a two-way link, but could work as a stepping stone to a better solution.

JIRA's link management is spot on.

**Represent a type of relation**
Each relation has a left-to-right and a right-to-left wording. For example:
```
[
    {leftToRight: "blocks", rightToLeft: "is blocked by"},
    {leftToRight: "depends on", rightToLeft: "is a dependency for"},
    {leftToRight: "is related to", rightToLeft: "is related to"}
]
```

Being able to customise available relation types is a plus, but not strictly necessary. Having "depends on", "blocks", "relates to" and "duplicates" is a good starting point.

**Represent a link**
```
{
    leftHandItem: <story/task/issue ID>,
    rightHandItem: <story/task/issue ID>,
    relation: <relation ID>
}
```

**UI**
A UI section "Links" below "Related tasks" for stories, and below the description for tasks and issues.

When creating a link, each relation type is represented twice in a dropdown, with leftToRight and rightToLeft wording. Depending on which one is chosen, the story we are currently editing becomes the leftHand or rightHand item.

If you agree with this rough spec, I might try to implement it. :smile:

**Comment by ncoghlan** [comment]
While I agree JIRA is a reasonable source of inspiration in terms of "What kinds of relationships do people want to model?", from a link representation perspective, the W3C's efforts with JSON-LD would be worth exploring for a link representation project that isn't bound by backwards compatibility requirements: http://json-ld.org/

**Comment by tabell** [comment]
+1 The dependency feature is quite important.

**Comment by zookz** [comment]
+1

**Comment by bushong1** [comment]
+1

**Comment by rubdos** [comment]
I just started testing taiga today, and I find it quite surprising not to find inter user stories and inter issue dependencies. Therefor: +1

I think it would really help; you could choose not to view cards that depend on others, making it easier to wade through the user stories...

**Comment by arno01** [comment]
+1 indeed. This would be a killer-feature :-)

**Comment by ph-One** [comment]
:+1:

**Comment by alexis-salo** [comment]
this is preventing us to use Taiga... so +1

**Comment by piribes** [comment]
+1

**Comment by mrichar1** [comment]
I'd also like to add a +1 to this.

I notice that user-stories can have tasks associated with them - is this a route to also associating issues?

**Comment by rubdos** [comment]
> The second is that any money donated will go towards a bounty fund. Starting Q2 2015 we will be engaging much more actively with our community to help further the development of Taiga, and we will use these donations to reward people working alongside us.

I have a new software development project upcoming. If the other party signs, I'm putting a 75 euro bounty on this, if that's possible.

**Comment by eaposner** [comment]
Hi. Perhaps this is something that can be built using Taiga.io new taigaTribe module now in testing mode: www.taigatribe.com ?

**Comment by timothylclark** [comment]
+1 - really, really need this feature. Linking US to issues is the first step; adding the ability to annotate those links would be even better.

**Comment by gabriel-sevecek** [comment]
+1

**Comment by macarena** [comment]
+1

**Comment by Nepoxx** [comment]
Github has a +1 feature now, please consider using it.

**Comment by BioPhoton** [comment]
Hi @superalex,

are there any news about the userstory - task relation?

**Comment by kranthitech** [comment]
I seriously think you guys should prioritize this.

Right now a lot of pivotal tracker users are unhappy that it doesn't support story dependency. If you can implement this and publish a few blog posts on how to use it, it will definitely become a compelling reason for people to shift from pivotal tracker to taiga.

**Comment by OldGlory747** [comment]
I think we should consider crowdfunding this.

**Comment by shiweiluanmicrovu** [comment]
+1

**Comment by mahtuag** [comment]
Has this changed with the Taiga 3.0 release ?

I'm quoting the newsletter:

>  Now you can set up user stories and tasks that and assign them to multiple projects. You'll be able to track hierarchical or related relationships between those user stories on different projects.

**Comment by bameda** [comment]
Yes, but read all the section and don't extract the phrase you want

> ### MULTIPROJECT EPICS
>
> Epics on Taiga are, well - Epic! Now you can set up user stories and tasks that and assign them to multiple projects. You'll be able to track hierarchical or related relationships between those user stories on different projects. This is particularly useful if your work requires planning across multiple projects, and you wish to set up user stories and tasks that don't belong to just one project.
>
> ![](https://blog.taiga.io/images/2016-10-02_changelog300/epics.gif)

**Comment by Xaviju** [comment]
@mahtuag Sort of. 
Its not exactly the same (here there are different versions). New Taiga release includes [Epics](https://tree.taiga.io/support/epics/what-is-an-epic/), Epics can contain User Stories from multiple projects and those User Stories contain tasks so its a hierarchical dependency.

**Comment by mahtuag** [comment]
@Xaviju thanks for the explanation.

So an epic can be used to create hierarchical dependencies between tasks. What about sequential dependencies, like A and B are prerequisites (or blockers) for C ?

@bameda I'm not very good with agile buzzwords. Took epic to mean [this](www.urbandictionary.com/define.php?term=epic) 😅

**Comment by Xaviju** [comment]
> So an epic can be used to create hierarchical dependencies between tasks

@mahtuag not exactly, between User Stories. Tasks are dependant on User Stories.

Sequential dependencies are not part of this release, but probably Epics will help a lot on achieving this behaviour.

**Comment by jmpace** [comment]
Linking issues to stories is a must have!

**Comment by CzEMC** [comment]
+1 The ability to simply link Issues to an User Story would be very useful in my opinion. It would offer easier organisation within a team to effortlessly check out each individual issue that is connected to the User Story. 

Even though an Issue could be copied to an User Story, it doesn't maintain its key characteristics and either it wouldn't be used or the Issue would be marked closed because further work on it would be completed using the User Story.

The implementation of linking each issue to a task in a User Story would be a good idea as well.

This would be similar to GitHub's issue & pull request referencing feature.

**Comment by jamesokbo** [comment]
+1 Tasks dependency

**Comment by javisantos** [comment]
+1

**Comment by LanderBeeuwsaert** [comment]
+1

**Comment by Portgas666** [comment]
+1

**Comment by sveinburne** [comment]
+1

**Comment by selatotal** [comment]
+1

**Comment by om-nishu-trantor** [comment]
Still not having Task dependencies ?
Really required, should have it in now.

**Comment by jim-hatama** [comment]
add this plz

**Comment by ppetermann** [comment]
is there any timeframe planed for this?

**Comment by milkeg** [comment]
Hello,
Any news on this? It would simplify our work :)
Thank you

**Comment by Boimb** [comment]
Hi,
3 years since this issue is open 😢 
Don't you still consider this feature as a must have ?

**Comment by sergeykorablin** [comment]
+1

**Comment by rbs-asimon** [comment]
Also thinking linking objects with other objects (I mean every type of object with every type of objects : a story with a story, a story with a bug ticket, a task with ...) is a must have :+1: 

These are the options JIRA uses to link objects :

![image](https://user-images.githubusercontent.com/3064469/42162681-f0877848-7dff-11e8-8456-607263f2ffbe.png)

**Comment by lastlink** [comment]
One way you could do is manually type out these dependencies w/ "#number of userstory" For now...

**Comment by dnk8n** [comment]
For now, could the tagging mechanism be used to specify links? For example

`#32:depends_on:#34`
`#34:depends_on:#36`

Could there be some logic built within the tagging system that calls a function ('depends_on'), when specified in this way, which would auto-tag both stories.

I know this is not ideal. But I thought that maybe it could be less work to get something rough out the door. In addition it might be nice to autohide these 'logic-tags' by default else things might look messy. Would also need a way to have two hyperlinks in the tag (leaving the depends_on section as plain text).

A further nice-to-have would be to generate a dependency graph view from these 'system/hidden tags'. Over a few iterations, this functionality could be abstracted away from the tagging system altogether.

Thoughts?


But for now, one could theoretically do this tagging manually to each story involved. Awful idea when you think about it like that. But with automation and abstraction it could be very powerful.

**Comment by Kourgiantakis** [comment]
FYI: Would have bought a subscription for our company but not being able to set dependencies sounds like a bad joke and is a total show stopper.

**Comment by zakkg3** [comment]
it's been 6 yrs since this issue was opened :D but I still will... +1

**Comment by kedare** [comment]
+1, a task management without the possibility to define relations and dependencies between them is kind of extremely limited :(

**Comment by arno01** [comment]
It is super sad to see its been 6 years and no dependencies.. 2nd company I am working at is instantly rejecting Taiga when I am bringing it up in the moments we are deciding moving to a new project management solution.

**Comment by lenkunz** [comment]
Isn't this a fundamental thing for project management?
How can this be left so low prioritized?

If this feature is up, I still need to manage this in sprint meeting.
Current workaround is custom attribute with markdown. Open the task in new tab to check if the story is ready.
If I know that task is ready by its own, it will be much faster.

**Comment by Kourgiantakis** [comment]
> 
> 
> Isn't this a fundamental thing for project management?
> How can this be left so low prioritized?
> 
> If this feature is up, I still need to manage this in sprint meeting.
> Current workaround is custom attribute with markdown. Open the task in new tab to check if the story is ready.
> If I have a way to know that the task is ready by its own, it will be much faster.

It IS fundamental and our company bought a self hosting subscription from ActiveCollab for this very reason. Sure, Taiga is cheaper but it lacks one of the most important feature.

**Comment by lpilz** [comment]
@NatachaMenjibar @superalex @jespino

This feature seems to be the most requested when sorting issues by comments and by upvotes. Since Taiga6 includes a large UI/UX overhaul (cf. taiga-front/#467), is this feature included / did you consider it / are you considering it a candidate for further development?

**Comment by yamila-moreno** [comment]
Hi #everybody thanks for the interest and keeping this issue alive. We're planning next steps and roadmap (including making a public and meaningful roadmap) and we're definetly taking this feedback into account.

Cheers!

**Comment by pihentagy** [comment]
By accepting a feature request I think this should not be closed.

**Comment by yamila-moreno** [comment]
Hi, we're taking this FR into account, but it won't be near soon, so I'm afraid we cannot give follow ups about this. Nonetheless, if you prefer it open, it's totally fine, of course :)

**Comment by pihentagy** [comment]
I just expected a link to a followup ticket where we can track further progress.

**Comment by xavierlaffargue** [comment]
@yamila-moreno so... are you going to do it or not?
